http://sourceware.org/ml/gdb-patches/2012-09/msg00632.html
Subject: [PATCH 2/4] Add a check to ensure that a type may fit into host memory


--MP_/PnL6l3LUsXWpZ/olqawWlzb
Content-Type: text/plain; charset=US-ASCII
Content-Transfer-Encoding: 7bit
Content-Disposition: inline

Hi,

This is part two of the bitpos expansion patch.  This implements checks
in some places in the code to ensure that a type size in ULONGEST is
small enough to fit into host memory.  Tested for regressions on x86_64
Fedora 16.

Regards,
Siddhesh

--MP_/PnL6l3LUsXWpZ/olqawWlzb
Content-Type: text/plain
Content-Transfer-Encoding: quoted-printable
Content-Disposition: attachment; filename=ChangeLog-ensure_sizet

gdb/ChangeLog

	* alpha-tdep.c (alpha_push_dummy_call) Check for underflow in
	SP.
	* cp-valprint (cp_print_value): Ensure BASECLASS fits into
	size_t.
	* dwarf2loc.c (read_pieced_value): Ensure that THIS_SIZE fits
	into size_t.
	(write_pieced_value): Likewise.
	* findcmd.c (parse_find_args): Ensure PATTERN_BUF_SIZE fits into
	size_t.
	* p-valprint (pascal_object_print_value): Ensure BASECLASS fits
	into size_t.
	* utils.c (ulongest_fits_host_or_error): New function to find if
	a ULONGEST number fits into size_t.
	* utils.h: Declare ulongest_fits_host_or_error.
	* valops.c (search_struct_method): Ensure BASECLASS fits into
	size_t.
	* value.c (allocate_value_lazy): Ensure TYPE fits into size_t.
	(allocate_value_contents): Likewise.
	(set_value_enclosing_type): Ensure NEW_ENCL_TYPE fits into
	size_t.
	* vax-tdep.c (vax_return_value): Ensure that TYPE fits into
	size_t.

--MP_/PnL6l3LUsXWpZ/olqawWlzb
Content-Type: text/x-patch
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment; filename=bitpos-ensure-size_t.patch

Index: gdb-7.6.90.20140127/gdb/alpha-tdep.c
===================================================================
--- gdb-7.6.90.20140127.orig/gdb/alpha-tdep.c	2014-02-06 18:18:51.005616676 +0100
+++ gdb-7.6.90.20140127/gdb/alpha-tdep.c	2014-02-06 18:18:53.671621349 +0100
@@ -414,6 +414,13 @@ alpha_push_dummy_call (struct gdbarch *g
     accumulate_size = 0;
   else
     accumulate_size -= sizeof(arg_reg_buffer);
+
+  /* Check for underflow.  */
+  if (sp - accumulate_size > sp)
+    error (_("Insufficient memory in GDB host for arguments, "
+	     "need %s bytes, but less than %s bytes available."),
+	   plongest (accumulate_size), plongest (CORE_ADDR_MAX - sp));
+
   sp -= accumulate_size;
 
   /* Keep sp aligned to a multiple of 16 as the ABI requires.  */
Index: gdb-7.6.90.20140127/gdb/cp-valprint.c
===================================================================
--- gdb-7.6.90.20140127.orig/gdb/cp-valprint.c	2014-02-06 18:18:51.006616677 +0100
+++ gdb-7.6.90.20140127/gdb/cp-valprint.c	2014-02-06 18:18:53.671621349 +0100
@@ -537,6 +537,8 @@ cp_print_value (struct type *type, struc
 		  gdb_byte *buf;
 		  struct cleanup *back_to;
 
+		  ulongest_fits_host_or_error (TYPE_LENGTH (baseclass));
+
 		  buf = xmalloc (TYPE_LENGTH (baseclass));
 		  back_to = make_cleanup (xfree, buf);
 
Index: gdb-7.6.90.20140127/gdb/dwarf2loc.c
===================================================================
--- gdb-7.6.90.20140127.orig/gdb/dwarf2loc.c	2014-02-06 18:18:51.007616678 +0100
+++ gdb-7.6.90.20140127/gdb/dwarf2loc.c	2014-02-06 18:18:53.672621293 +0100
@@ -1821,6 +1821,8 @@ read_pieced_value (struct value *v)
 
       this_size = (this_size_bits + source_offset_bits % 8 + 7) / 8;
       source_offset = source_offset_bits / 8;
+      ulongest_fits_host_or_error (this_size);
+
       if (buffer_size < this_size)
 	{
 	  buffer_size = this_size;
@@ -2012,6 +2014,7 @@ write_pieced_value (struct value *to, st
 	}
       else
 	{
+	  ulongest_fits_host_or_error (this_size);
 	  if (buffer_size < this_size)
 	    {
 	      buffer_size = this_size;
Index: gdb-7.6.90.20140127/gdb/findcmd.c
===================================================================
--- gdb-7.6.90.20140127.orig/gdb/findcmd.c	2014-02-06 18:18:51.008616679 +0100
+++ gdb-7.6.90.20140127/gdb/findcmd.c	2014-02-06 18:18:53.673621239 +0100
@@ -185,6 +185,7 @@ parse_find_args (char *args, ULONGEST *m
 	  size_t current_offset = pattern_buf_end - pattern_buf;
 
 	  pattern_buf_size = pattern_buf_size_need * 2;
+	  ulongest_fits_host_or_error (pattern_buf_size);
 	  pattern_buf = xrealloc (pattern_buf, pattern_buf_size);
 	  pattern_buf_end = pattern_buf + current_offset;
 	}
Index: gdb-7.6.90.20140127/gdb/p-valprint.c
===================================================================
--- gdb-7.6.90.20140127.orig/gdb/p-valprint.c	2014-02-06 18:18:51.008616679 +0100
+++ gdb-7.6.90.20140127/gdb/p-valprint.c	2014-02-06 18:18:53.673621239 +0100
@@ -798,6 +798,7 @@ pascal_object_print_value (struct type *
 	      gdb_byte *buf;
 	      struct cleanup *back_to;
 
+	      ulongest_fits_host_or_error (TYPE_LENGTH (baseclass));
 	      buf = xmalloc (TYPE_LENGTH (baseclass));
 	      back_to = make_cleanup (xfree, buf);
 
Index: gdb-7.6.90.20140127/gdb/utils.c
===================================================================
--- gdb-7.6.90.20140127.orig/gdb/utils.c	2014-02-06 18:18:51.008616679 +0100
+++ gdb-7.6.90.20140127/gdb/utils.c	2014-02-06 18:18:53.674621186 +0100
@@ -3169,6 +3169,18 @@ host_address_to_string (const void *addr
   return str;
 }
 
+/* Ensure that the input NUM is not larger than the maximum capacity of the
+   host system.  We choose SIZE_MAX / 8 as a conservative estimate of the size
+   of a resource that a system may allocate.  */
+void
+ulongest_fits_host_or_error (ULONGEST num)
+{
+  if (num > SIZE_MAX / 8)
+    error (_("Insufficient memory in host GDB for object of size %s bytes, "
+	     "maximum allowed %s bytes."), pulongest (num),
+	   pulongest (SIZE_MAX / 8));
+}
+
 char *
 gdb_realpath (const char *filename)
 {
Index: gdb-7.6.90.20140127/gdb/valops.c
===================================================================
--- gdb-7.6.90.20140127.orig/gdb/valops.c	2014-02-06 18:18:51.009616680 +0100
+++ gdb-7.6.90.20140127/gdb/valops.c	2014-02-06 18:18:53.675621135 +0100
@@ -2184,6 +2184,7 @@ search_struct_method (const char *name,
 	      struct cleanup *back_to;
 	      CORE_ADDR address;
 
+	      ulongest_fits_host_or_error (TYPE_LENGTH (baseclass));
 	      tmp = xmalloc (TYPE_LENGTH (baseclass));
 	      back_to = make_cleanup (xfree, tmp);
 	      address = value_address (*arg1p);
Index: gdb-7.6.90.20140127/gdb/value.c
===================================================================
--- gdb-7.6.90.20140127.orig/gdb/value.c	2014-02-06 18:18:51.010616681 +0100
+++ gdb-7.6.90.20140127/gdb/value.c	2014-02-06 18:19:10.261637398 +0100
@@ -822,6 +822,7 @@ allocate_value_lazy (struct type *type)
      description correctly.  */
   check_typedef (type);
 
+  ulongest_fits_host_or_error (TYPE_LENGTH (type));
   val = (struct value *) xzalloc (sizeof (struct value));
   val->contents = NULL;
   val->next = all_values;
@@ -853,6 +854,8 @@ allocate_value_lazy (struct type *type)
 static void
 allocate_value_contents (struct value *val)
 {
+  ulongest_fits_host_or_error (TYPE_LENGTH (val->enclosing_type));
+
   if (!val->contents)
     val->contents = (gdb_byte *) xzalloc (TYPE_LENGTH (val->enclosing_type));
 }
@@ -2854,8 +2857,12 @@ void
 set_value_enclosing_type (struct value *val, struct type *new_encl_type)
 {
   if (TYPE_LENGTH (new_encl_type) > TYPE_LENGTH (value_enclosing_type (val))) 
-    val->contents =
-      (gdb_byte *) xrealloc (val->contents, TYPE_LENGTH (new_encl_type));
+    {
+      ulongest_fits_host_or_error (TYPE_LENGTH (new_encl_type));
+
+      val->contents =
+	(gdb_byte *) xrealloc (val->contents, TYPE_LENGTH (new_encl_type));
+    }
 
   val->enclosing_type = new_encl_type;
 }
Index: gdb-7.6.90.20140127/gdb/vax-tdep.c
===================================================================
--- gdb-7.6.90.20140127.orig/gdb/vax-tdep.c	2014-02-06 18:18:51.370617068 +0100
+++ gdb-7.6.90.20140127/gdb/vax-tdep.c	2014-02-06 18:18:53.676621086 +0100
@@ -223,6 +223,7 @@ vax_return_value (struct gdbarch *gdbarc
 	  ULONGEST addr;
 
 	  regcache_raw_read_unsigned (regcache, VAX_R0_REGNUM, &addr);
+	  ulongest_fits_host_or_error (TYPE_LENGTH (type));
 	  read_memory (addr, readbuf, len);
 	}
 
Index: gdb-7.6.90.20140127/gdb/defs.h
===================================================================
--- gdb-7.6.90.20140127.orig/gdb/defs.h	2014-02-06 18:18:51.370617068 +0100
+++ gdb-7.6.90.20140127/gdb/defs.h	2014-02-06 18:18:53.677621038 +0100
@@ -768,4 +768,6 @@ enum block_enum
 
 #include "utils.h"
 
+extern void ulongest_fits_host_or_error (ULONGEST num);
+
 #endif /* #ifndef DEFS_H */
